jerry-core/ecma/builtin-objects/ecma-builtin-number-prototype.c

```diff
@@ -651,89 +613,106 @@ static ecma_value_t
   }
 
-  /* Get the parameters of the number if non-zero. */
-  lit_utf8_byte_t digits[ECMA_MAX_CHARS_IN_STRINGIFIED_NUMBER];
-  lit_utf8_size_t num_digits;
-  int32_t exponent;
+  /* 2. */
+  if (arg_num < 0 || arg_num > 20)
+  {
+    return ecma_raise_range_error (ECMA_ERR_MSG ("Fraction digits must be between 0 and 20."));
+  }
 
-  if (!ecma_number_is_zero (this_num))
+  /* 4. */
+  if (ecma_number_is_nan (this_num))
   {
-    num_digits = ecma_number_to_binary_floating_point_number (this_num, digits, &exponent);
-    JERRY_ASSERT (exponent >= 0);
+    return ecma_make_magic_string_value (LIT_MAGIC_STRING_NAN);
   }
-  else
+
+  /* 5. */
+  ecma_stringbuilder_t builder = ecma_stringbuilder_create ();
+  int32_t fraction_digits = ecma_number_to_int32 (arg_num);
+  bool is_zero = ecma_number_is_zero (this_num);
+
+  /* 6. */
+  if (ecma_number_is_negative (this_num))
   {
-    for (int32_t i = 0; i <= frac_digits; i++)
+    if (!is_zero)
     {
-      digits[i] = '0';
+      ecma_stringbuilder_append_char (&builder, LIT_CHAR_MINUS);
     }
-    num_digits = (lit_utf8_size_t) frac_digits + 1;
-    exponent = 1;
+
+    this_num *= -1;
   }
 
-  /* 7. */
-  if (exponent > 21)
+  /* Handle zero separately */
+  if (is_zero)
   {
-    return ecma_builtin_number_prototype_object_to_string (this_num, NULL, 0);
-  }
+    ecma_stringbuilder_append_char (&builder, LIT_CHAR_0);
 
-  /* 8. */
-  num_digits = ecma_builtin_number_prototype_helper_round (digits,
-                                                           num_digits + (lit_utf8_size_t) exponent,
-                                                           exponent + frac_digits,
-                                                           &exponent,
-                                                           ecma_number_is_zero (this_num));
+    if (fraction_digits > 0)
+    {
+      ecma_stringbuilder_append_char (&builder, LIT_CHAR_DOT);
+    }
 
-  /* Buffer that is used to construct the string. */
-  int buffer_size = (exponent > 0) ? exponent + frac_digits + 2 : frac_digits + 3;
+    for (int32_t i = 0; i < fraction_digits; i++)
+    {
+      ecma_stringbuilder_append_char (&builder, LIT_CHAR_0);
+    }
 
-  if (is_negative)
-  {
-    buffer_size++;
+    return ecma_make_string_value (ecma_stringbuilder_finalize (&builder));
   }
 
-  JERRY_ASSERT (buffer_size > 0);
+  /* Handle infinity separately */
+  if (ecma_number_is_infinity (this_num))
+  {
+    ecma_stringbuilder_append_magic (&builder, LIT_MAGIC_STRING_INFINITY_UL);
+    return ecma_make_string_value (ecma_stringbuilder_finalize (&builder));
+  }
 
-  ecma_value_t ret_value;
-  JMEM_DEFINE_LOCAL_ARRAY (buff, buffer_size, lit_utf8_byte_t);
+  /* Get the parameters of the number */
+  lit_utf8_byte_t digits[ECMA_MAX_CHARS_IN_STRINGIFIED_NUMBER];
+  lit_utf8_size_t num_of_digits;
+  int32_t exponent;
 
-  lit_utf8_byte_t *p = buff;
+  num_of_digits = ecma_number_to_binary_floating_point_number (this_num, digits, &exponent);
```

Comment:
I think this function should be removed as well. The other functions all use `ecma_number_to_decimal` to convert the number to digits, we should use that here as well.

The names of some of the variables should also be revised, they seem to store different values than what they imply. For example currently calling `ecma_number_to_binary_floating_point_number ` on the number `5` will result in `num_digits = 0, digits = "5", exponenet = 1`, which makes no sense at all.

After switching to `ecma_number_to_decimal` we should double check the variables and name them appropriately. Maybe even `ecma_number_to_decimal` should be renamed/updated a bit so that it's purpose and the meaning of the returned values are clearer.